packages/frontend/core/src/modules/cmdk/entities/quick-search-manager.ts

```diff
@@ -0,0 +1,192 @@
+    let recentPages = this.getRecentDocIds();
+    recentPages = recentPages.filter(id => id !== pageId);
+    if (recentPages.length >= RECENT_PAGES_LIMIT) {
+      recentPages.pop();
+    }
+    recentPages.unshift(pageId);
+    this.localState.set(RECENT_PAGES_KEY, recentPages);
+  }
+
+  getRecentDocs() {
+    const docs = this.docsService.list.docs$.value;
+    return this.getRecentDocIds()
+      .map(id => docs.find(doc => doc.id === id))
+      .filter((d): d is DocRecord => !!d);
+  }
+
+  showCMDK() {
+    return this.show('cmdk');
+  }
+
+  hide() {
+    return this.show(null);
+  }
+
+  toggle() {
+    return this.show$.value ? this.hide() : this.showCMDK();
+  }
+
+  search(query?: string) {
+    const counter = this.show('search-callback');
+    this.setQuery(query || '');
+    return new Promise<SearchCallbackResult | null>(resolve => {
+      const sub = race(
+        this.show$.pipe(skip(1)),
+        this._searchCallbackResult$
+      ).subscribe(result => {
+        if (
+          result &&
+          typeof result === 'object' &&
+          counter === this.showCounter
+        ) {
+          resolve(result);
+        } else {
+          resolve(null);
+        }
+        sub.unsubscribe();
+      });
+    });
+  }
+
+  setSearchCallbackResult(result: SearchCallbackResult) {
+    this._searchCallbackResult$.next(result);
+  }
+
+  getSearchedDocs(query: string) {
+    const searchResults = this.workspaceService.workspace.docCollection.search(
+      query
+    ) as unknown as Map<
+      string,
+      {
+        space: string;
+        content: string;
+      }
+    >;
+    // make sure we don't add the same page multiple times
+    const added = new Set<string>();
+    const docs = this.docsService.list.docs$.value;
+    const searchedDocs: {
+      doc: DocRecord;
+      blockId: string;
+      content?: string;
+      source: 'search' | 'link-ref';
+    }[] = Array.from(searchResults.entries())
+      .map(([blockId, { space, content }]) => {
+        const doc = docs.find(doc => doc.id === space && !added.has(doc.id));
+        if (!doc) return null;
+        added.add(doc.id);
+
+        return {
+          doc,
+          blockId,
+          content,
+          source: 'search' as const,
+        };
+      })
+      .filter((res): res is NonNullable<typeof res> => !!res);
+
+    const maybeRefLink = resolveLinkToDoc(query);
+
+    if (maybeRefLink) {
+      const doc = this.docsService.list.docs$.value.find(
+        doc => doc.id === maybeRefLink.docId
+      );
+      if (doc) {
+        searchedDocs.push({
+          doc,
+          blockId: maybeRefLink.blockId,
+          source: 'link-ref',
+        });
+      }
+    }
+
+    return searchedDocs;
+  }
+
+  private getRecentDocIds() {
+    return this.localState.get(RECENT_PAGES_KEY) || EMPTY_ARRAY;
+  }
+}
+
+declare module '@toeverything/infra' {
+  interface WorkspaceLocalState {
+    get(key: typeof RECENT_PAGES_KEY): string[] | null;
```

Comment:
You can directly use the generic parameters on WorkspaceLocalState.

Reply:
I think overriding `get`, `set` etc will provide better type hints universally even outside of the current module though.